### PR TITLE
AppInsights telemetry

### PR DESCRIPTION
**Pull request by EasyRhinoMSFT**

### CODE REVIEW

File: src/Sarif.Viewer.VisualStudio/App.config [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <startup>
+    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5.2" />
+  </startup>
+  <appSettings>
+    <add key="TelemetryInstrumentationKey_Debug" value="80a368bb-f285-46eb-bb20-3e424593ec56"/>
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/App.config]: >80a368bb-f285-46eb-bb20-3e424593ec56 [](start = 56, length = 36)

Do we want the same GUID for debug vs. release? The nice thing about two distinct guids is that we can debug/test against one and not compromise release telemetry.

Comment (EasyRhinoMSFT) [reply src/Sarif.Viewer.VisualStudio/App.config]: Yes that is the one outstanding to-do. I can't create a new instance so I'll need you to do that for me.

File: src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.ApplicationInsights;
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: >Microsoft.ApplicationInsights; [](start = 6, length = 30)

using System directives should come first. you might need to change your VS settings.

File: src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+using System.Linq;
+
+namespace Microsoft.Sarif.Viewer
+{
+    /// <summary>
+    /// Wrapper class for Application Insights client.
+    /// </summary>
+    internal static class TelemetryProvider
+    {
+        private static TelemetryClient appInsightsClient;
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: >appInsightsClient [](start = 39, length = 17)

please use s_ as a prefix for all private static members.

File: src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        /// </summary>
+        /// <param name="toolFormat">The tool format of the menu item.</param>
+        public static void WriteMenuCommandEvent(string toolFormat)
+        {
+            if (string.IsNullOrWhiteSpace(toolFormat))
+            {
+                throw new ArgumentNullException(nameof(toolFormat));
+            }
+
+            WriteEvent(TelemetryEvent.LogFileOpenedByMenuCommand, "Format".KeyWithValue(toolFormat == ToolFormat.None ? "SARIF" : toolFormat));
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: >ToolFormat.None [](start = 102, length = 15)

Is ToolFormat.None really a valid value to apss here? Or are we masking some other issue if this code path is exercised?

**Comment by michaelcfanning** [comment]
@komal436 @jeff-bacon

File: src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs (hunk 4 above)
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: >"Format" [](start = 66, length = 8)

what's with this pattern? is there an extension method somewhere that adds KeyWithValue to the string type? 

I'm not sure I like this., the issue is that we're adding a bolt-on method to a primitive type that's used pervasively in scenarios where KeyWithValue doesn't make sense to apply. 

I'll keep my eyes open for the extension method. if it's defined within a narrow namespace, could be ok.

Comment (EasyRhinoMSFT) [reply src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: Yes it's in the new file String.extensions.cs

Comment (EasyRhinoMSFT) [reply src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: Yeah if you look at OpenLogFileCommands.cs line 168, you see that the tool format is left at None for sarif files.

File: src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs (hunk 2 above)
Comment (EasyRhinoMSFT) [reply src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: Cool, didn't know about that option.

File: src/Sarif.Viewer.VisualStudio/String.extensions.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Sarif.Viewer
+{
+    static class StringExtensions
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/String.extensions.cs]: >static [](start = 4, length = 6)

please provide accessibility modifier (internal) for this class

**Comment by michaelcfanning** [comment]
> // Copyright (c) Microsoft. All rights reserved.

Can you please name this file StringExtensions.cs, which matches the type name?

---
Refers to: src/Sarif.Viewer.VisualStudio/String.extensions.cs:1 in e467a26. [](commit_id = e467a2658377cb292a9da5772af822a4893192f6, deletion_comment = False)

File: src/Sarif.Viewer.VisualStudio/String.extensions.cs (hunk 5 above)
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/String.extensions.cs]: >StringExtensions [](start = 17, length = 16)

Do me a favor, zap this class, actually, and go add a static helper to TelemetryEvent or TelemetryProvider that does this construction.

problems here (all small)
1) new type with a single method
2) extension method that applies a method to a commonly used type where the method is only applicable in a small number of scenarios where that type is used

alternately, we could have put this extension method in a sub-namespace that would restrict use in contexts where we opt into the telemetry. that would entail putting the telemetry types in their own namespace.

File: src/Sarif.Viewer.VisualStudio/String.extensions.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        public static KeyValuePair<string, string> KeyWithValue(this string key, string value)
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/String.extensions.cs]: >KeyValuePair [](start = 22, length = 12)

where are the unit tests for this admittedly simple method?

pretty easy to test. :)

File: src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Sarif.Viewer
+{
+    /// <summary>
+    /// Wrapper class for Application Insights client.
+    /// </summary>
+    internal static class TelemetryProvider
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio/TelemetryProvider.cs]: >TelemetryProvider [](start = 26, length = 17)

tests for this class and other?

you can defer this work. suggest opening issues for it. we do want to get the telemetry into the viewer release if possible

File: src/Sarif.Viewer.VisualStudio/String.extensions.cs (hunk 5 above)
Comment (EasyRhinoMSFT) [reply src/Sarif.Viewer.VisualStudio/String.extensions.cs]: I'd rather just yank it and do KVP construction directly.

**Comment by michaelcfanning** [comment]
Chris, have you verified the telemetry is functioning? 

---
In reply to: [353115509](https://github.com/Microsoft/sarif-sdk/pull/729#issuecomment-353115509) [](ancestors = 353115509)

File: src/Sarif.Viewer.VisualStudio/App.config (hunk 1 above)
Comment (michaelcfanning) [reply src/Sarif.Viewer.VisualStudio/App.config]: ok, open an issue on it for chris faucon

---
In reply to: [158073364](https://github.com/Microsoft/sarif-sdk/pull/729#discussion_r158073364) [](ancestors = 158073364)

**Pull request merged**